### ISSUE TITLE
(new Entity()).IsAlive() is maybe true

### ISSUE THREAD

**Issue by RoyAwesome** [genaray/Arch#53]
This a surprising behavior.

```cs
{
   World w = World.Create();
   bool e_alive_1 = (new Entity()).IsAlive(); //Returns false
   w.Create();
   bool e_alive_2 = (new Entity()).IsAlive(); //Returns true
}
```

this should probably return false in both situations.  This leads to surprising results if you are trying to store an entity in a component and want to check if the parameter has been initialized or not.

**Comment by genaray** [comment]
Thanks ! 

Is this pseudocode ? 
The entity constructor is internal and therefore can and should not be used like this.

That's why I'm not sure right now if it's a real problem or just unlawful access ^^

**Comment by RoyAwesome** [comment]
No, it's real code I wrote for testing.  You can default construct structs in C# without a public constructor.

**Comment by genaray** [comment]
> No, it's real code I wrote for testing to figure out why IsAlive wasn't behaving as I expected. You can default construct structs in C# without a public constructor.

Hmmm... I see, I strongly suspect that I might have to do the complete entity internal. With the constructor the user should not play around because such behavior is undefined. 

If you want to reference entities, Entity.Null is used. 

```csharp
cmpWithEntityReference.Entity = Entity.Null; // Do not do : new Entity();
if(cmpWithEntityRefernce.Entity != Entity.Null && cmpWithEntityRefernce.Entity.IsAlive())
   cmpWithEntityReference.Entity.Set(...);
```

**Comment by genaray** [comment]
https://github.com/genaray/Arch/releases/tag/1.1.5 now has a feature called "EntityReference" which is used for... well referencing entities.

That should clarify stuff a bit, in the future i will edit the docs to make clear that `entity.IsAlive()` is only used to check if the entity id exists in the world.

**Comment by RoyAwesome** [comment]
if I have some struct

```cs
struct MyComponent
{
   public EntityReference Ref;
}
```

and I do some

```cs
var Foo = World.Create(new MyComponent());

Foo.Get<MyComponent>().Ref.IsAlive()
```

would that reasonably return false with this new object?

**Comment by genaray** [comment]
> if I have some struct
> 
> ```cs
> struct MyComponent
> {
>    public EntityReference Ref;
> }
> ```
> 
> and I do some
> 
> ```cs
> var Foo = World.Create(new MyComponent());
> 
> Foo.Get<MyComponent>().Ref.IsAlive()
> ```
> 
> would that reasonably return false with this new object?

Exactly, with the newest package 1.1.6 this would return false since there was no entity yet referenced.
However once you assign a valid reference, it will return true as long as that entity id exists in the world and its version has not changed. So once the entity was NOT replaced/recyled `IsAlive` returns true and once it was replaced/recycled it will return false.

```csharp

var reference = new EntityReference(); // Or EntityReference.Null
reference.IsAlive() // False

reference = entity.Reference();
reference.IsAlive() // True
world.Destroy(entity);

reference.IsAlive() // Is False 
```

**Comment by genaray** [comment]
Completly fixed in [addd231](https://github.com/genaray/Arch/commit/addd231b92b24c9d20b168c788c83ea0115ee85e).
